crates/bevy_ecs/src/entity/mod.rs

```diff
@@ -115,18 +115,55 @@ type IdCursor = isize;
 /// [`EntityCommands`]: crate::system::EntityCommands
 /// [`Query::get`]: crate::system::Query::get
 /// [`World`]: crate::world::World
-#[derive(Clone, Copy, Eq, Ord, PartialOrd)]
+#[derive(Clone, Copy, Eq)]
+// Alignment repr necessary to allow LLVM to better output
+// optimised codegen for `to_bits`, `PartialEq` and `Ord`.
+// See <https://github.com/rust-lang/rust/issues/106107>
+#[repr(align(8))]
 pub struct Entity {
-    generation: u32,
+    // The field order below is necessary for rust to
+    // optimise better how it handles `Entity`
     index: u32,
+    generation: u32,
```

Comment:
Going to second this, since the conversion in `to_bits` is order dependent. For completeness, this should incorporate #3788's changes as well, since the ordering also relies on the endianness of the platform being targeted.

Reply:
I added both `#[repr(C, align(8))]` as well as the endian layout from #3788, so this should have bases covered on both parts, along with a comment explaining why this is so.